### PR TITLE
feat(oidc): make it possible to validate tokens against provider host…

### PR DESCRIPTION
**Pull request by fbartels**
…ed jwks document

Fixes: https://github.com/apache/apisix/issues/5425

Signed-off-by: Felix Bartels <felix@9wd.eu>

### What this PR does / why we need it:

At present one either needs to hard code the public key of the oidc provider in the configuration or the oidc provider needs to provide an introspection endpoint. There is however also the possibility to that tokens can be verified based on jwks endpoint of the provider. 

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [ ] Have you added corresponding test cases?
* [x] Have you modified the corresponding document?
* [x] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

### CODE REVIEW

**Comment by endersonmenezes** [comment]
This correction works for me! Thx :)

**Comment by juzhiyuan** [comment]
Hi @fbartels @endersonmenezes, thanks for your contribution and keep watching on https://github.com/apache/apisix 😄 I have requested one of our reviewers to review this PR.

**Review by spacewander — COMMENTED** [review]
Could you add two tests to validate this feature?
1. fail to verify the jwt token
2. verify the jwt token

You can read https://github.com/apache/apisix/blob/master/docs/en/latest/internal/testing-framework.md to know how to test.

File: t/plugin/openid-connect.t [hunk 1]
```diff
@@ -1487,3 +1487,140 @@ GET /t
...
+    location /t {
+        content_by_lua_block {
+            local t = require("lib.test_admin").test
+            local code, body = t('/apisix/admin/routes/1',
+                 ngx.HTTP_PUT,
+                 [[{ "plugins": {
+                            "openid-connect": {
+                                "client_id": "kbyuFDidLLm280LIwVFiazOqjO3ty8KH",
+                                "client_secret": "60Op4HFM0I8ajz0WdiStAbziZ-VFQttXuxixHHs2R7r7-CW8GR79l-mmLqMhc-Sa",
+                                "discovery": "https://samples.auth0.com/.well-known/openid-configuration",
```
Comment (tzssangglass) [inline t/plugin/openid-connect.t:1504]: @spacewander  Is it ok to rely on external services in a test case?  I'm a bit worried. Although I have found that some of our test cases rely on external services like httpbin.org.

Comment (tzssangglass) [reply t/plugin/openid-connect.t:1504]: @fbartels  Will `https://samples.auth0.com/.well-known/openid-configuration`  be valid for a long time (at least 2 years)?

Comment (fbartels) [reply t/plugin/openid-connect.t:1504]: @tzssangglass that is a good question. I could not really find much information about this domain in the Auth0 documentation.

I mostly copied the tests based on the public_key tests that were already in the testsuite. Other tests besides them were also using Auth0.

Comment (tzssangglass) [reply t/plugin/openid-connect.t:1504]: I get, my mistake😓, let's keep this.

Comment (fbartels) [reply t/plugin/openid-connect.t:1504]: No problem at all

File: t/plugin/openid-connect.t [hunk 2]
```diff
@@ -1487,3 +1487,140 @@ GET /t
...
+--- request
+GET /t
+--- response_body
+passed
+--- no_error_log
+[error]
+
+
+
+=== TEST 26: Access route with valid token.
```
Comment (spacewander) [inline t/plugin/openid-connect.t:1567]: This test fails with error "openid-connect.lua:268: phase_func(): OIDC introspection failed: JWT doesn't specify kid but the keystore contains multiple RSA keys"

Comment (fbartels) [reply t/plugin/openid-connect.t:1567]: Ahm. yes indeed. I checked the tokens that are already part of the test suite, but sadly none of them have a kid (key id) in its header. I have searched on the auth0 webpage for matching example tokens, but could not find one.

Comment (moonming) [reply t/plugin/openid-connect.t:1567]: @tzssangglass Can you provide some help? thanks.

Comment (tzssangglass) [reply t/plugin/openid-connect.t:1567]: ok, I will.

Comment (moonming) [reply t/plugin/openid-connect.t:1567]: any update？ @tzssangglass

Comment (tzssangglass) [reply t/plugin/openid-connect.t:1567]: will do this in the next two days.

**Comment by tzssangglass** [comment]
I tried, but I could not generate a JWT to pass the auth0 check. I gave up.

**Comment by juzhiyuan** [comment]
> I tried, but I could not generate a JWT to pass the auth0 check. I gave up.

May I know the reason we cannot do this? 😄 Maybe we could ask help from @spacewander @shuaijinchao ?

**Comment by tzssangglass** [comment]
> > I tried, but I could not generate a JWT to pass the auth0 check. I gave up.
> 
> May I know the reason we cannot do this? 😄 Maybe we could ask help from @spacewander @shuaijinchao ?

if we use `https://samples.auth0.com/.well-known/openid-configuration` as auth endpoint, and we want to test jwks, the public key for the jwks authentication method of JWT is stored in the `https://samples.auth0.com/.well-known/jwks.json`, I can't find the corresponding private key to generate a JWT to pass authentication.

hi, @sshniro , do you know anything about this?

**Comment by sshniro** [comment]
Hi @tzssangglass I'll check on this and get back tomorrow. However, I do recall that Auth0 had limitations when writing an extensive test suite (Eg- it was missing the introspection endpoint at the time of writing). This is one of the reasons why I used Keycloak as an IDP to write some test cases. And I can verify the token generated by Keycloak provides this KID, so if we cannot generate a token via Auth 0, then we can switch the test case to use Keycloak.

Screenshot 2021-11-23 at 08 37 49

**Comment by tzssangglass** [comment]
> then we can switch the test case to use Keycloak.

I like this way, but I'm not familiar with Keycloak, so please

**Comment by starsz** [comment]
> Hi @tzssangglass I'll check on this and get back tomorrow. However, I do recall that Auth0 had limitations when writing an extensive test suite (Eg- it was missing the introspection endpoint at the time of writing). This is one of the reasons why I used Keycloak as an IDP to write some test cases. And I can verify the token generated by Keycloak provides this KID, so if we cannot generate a token via Auth 0, then we can switch the test case to use Keycloak.
> 
> Screenshot 2021-11-23 at 08 37 49

Yes. I think we should use keycloak to test.`https://samples.auth0.com/` maybe not an official test website.

**Comment by GreenShadeZhang** [comment]
This is a very good feature, I want to say when you can merge it, I can't wait to experience it.

**Comment by moonming** [comment]
@starsz please take a look, thx

File: t/plugin/openid-connect.t (hunk 2 above)
Comment (starsz) [reply t/plugin/openid-connect.t:1567]: @fbartels , we can use `keycloak` instead of `samples.auth0` in the test.

**Review by starsz — COMMENTED** [review]
Hi @

File: t/plugin/openid-connect.t (hunk 2 above)
Comment (moonming) [reply t/plugin/openid-connect.t:1567]: @starsz, it will be great if you can show an example for you suggestion

Comment (starsz) [reply t/plugin/openid-connect.t:1567]: OK. @moonming Let me have a try.

**Comment by fbartels** [comment]
Author of the PR here, just to avoid confusion. I have deployed the above fix already internally and its working there for me. From what I can see here merging is currently blocked because of currently used testing infrastructure (Tokens and Endpoints from Auth0) missing properties in the Token and therefore tests failing. I personally am not working on the issues with the testing infrastructure.

**Comment by juzhiyuan** [comment]
> Author of the PR here, just to avoid confusion. I have deployed the above fix already internally and its working there for me. From what I can see here merging is currently blocked because of currently used testing infrastructure (Tokens and Endpoints from Auth0) missing properties in the Token and therefore tests failing. I personally am not working on the issues with the testing infrastructure.

Thanks for your update, and when you're free, welcome to continue on this PR 😄

**Comment by starsz** [comment]
Hi, @fbartels thanks for your contribution.
Now I will help you to finish this PR as co-author.

File: apisix/plugins/openid-connect.lua [hunk 3]
```diff
@@ -196,6 +196,7 @@ local function introspect(ctx, conf)
         end
 
         -- Token successfully validated.
+        core.log.debug("token validate successfully by ", conf.public_key and "public_key" or "jwk")
```
Comment (tzssangglass) [inline apisix/plugins/openid-connect.lua:199]:
```suggestion
        core.log.debug("token validate successfully by ", (conf.public_key and "public_key") or (conf.use_jwks and "jwk"))
```

is better?

Comment (starsz) [reply apisix/plugins/openid-connect.lua:199]: Good idea. Fix it.

**Review by tzssangglass — APPROVED** [review]

**Review by tokers — APPROVED** [review]

**Review by spacewander — APPROVED** [review]

**Pull request merged**